File: Ruby/021_merge_two_sorted_lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def merge_two_lists(l1, l2)
    l3 = ListNode.new(0)
    buf = l3

    while l1 && l2

        if l1.val <= l2.val
            l3.next = l1
            l1 = l1.next
        else
            l3.next = l2
            l2 = l2.next
        end

        l3 = l3.next
    end

    l3.next = l1 || l2
    buf.next

end

File: Python/021_merge_two_sorted_lists.py
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = cur = ListNode(0)
        while list1 and list2:
            if list1.val < list2.val:
                cur.next = list1
                list1 = list1.next
            else:
                cur.next = list2
                list2 = list2.next
            cur = cur.next
        cur.next = list1 or list2
        return dummy.next


# # If we can use function
# a = [1, 3, 4]
# b = [1, 2, 4]

# a.extend(b)
# a.sort()
# print(a)

# # or

# c = a+b
# c.sort()
# print(c)
